src/main/java/org/truffleruby/language/arguments/ReadArgumentsNode.java

```diff
@@ -0,0 +1,523 @@
+                            getContext(),
+                            coreExceptions().argumentError(given, dynamicArity.getRequired(), this));
+                }
+
+                final RubyHash keywordArguments = readUserKeywordsHashNode.execute(frame, descriptor);
+
+                if (!dynamicArity.hasKeywordsRest() && keywordArguments != null) {
+                    if (checkKeywordArgumentsNode == null) {
+                        CompilerDirectives.transferToInterpreterAndInvalidate();
+                        checkKeywordArgumentsNode = insert(new CheckKeywordArgumentsNode(getLanguage(), dynamicArity));
+                    }
+
+                    hashes.eachEntry(
+                            keywordArguments.store,
+                            keywordArguments,
+                            checkKeywordArgumentsNode,
+                            argumentsCount);
+                }
+            }
+        } else {
+            if (!dynamicArity.check(given)) {
+                checkArityError(dynamicArity, given, this);
+            }
+        }
+    }
+
+    public static void checkArityError(Arity arity, int given, Node currentNode) {
+        final RubyContext context = RubyContext.get(currentNode);
+        if (arity.hasRest()) {
+            throw new RaiseException(
+                    context,
+                    context.getCoreExceptions().argumentErrorPlus(given, arity.getRequired(), currentNode));
+        } else if (arity.getOptional() > 0) {
+            throw new RaiseException(
+                    context,
+                    context.getCoreExceptions().argumentError(
+                            given,
+                            arity.getRequired(),
+                            arity.getOptional(),
+                            currentNode));
+        } else {
+            throw new RaiseException(
+                    context,
+                    context.getCoreExceptions().argumentError(given, arity.getRequired(), currentNode));
+        }
+    }
+
+    private final BranchProfile expandProfile = BranchProfile.create();
+
+    private void expandKeywordHash(VirtualFrame frame, NonEmptyKeywordDescriptor descriptor) {
+        if (needsExpandedHash) {
+            if (!RubyGuards.isRubyHash(RubyArguments.getLastArgument(frame))) {
+                expandProfile.enter();
+                RubyArguments.setArgument(
+                        frame,
+                        descriptor.getHashIndex(),
+                        expandKeywordArgumentsNode.execute(
+                                Arrays.copyOf(frame.getArguments(), frame.getArguments().length), descriptor,
+                                RubyArguments.RUNTIME_ARGUMENT_COUNT));
+            }
+        }
+    }
+
+    @ExplodeLoop
+    private void clearFrameSlots(VirtualFrame frame) {
+        for (FrameSlot frameSlot : keywordParameterFrameSlots) {
+            frame.setObject(frameSlot, ReadArgumentsNode.MISSING);
```

Comment:
I wonder if this might be problematic if e.g. a `setInt` is done later, I'll check with Truffle/frame people.

Reply:
It's probably fine given that it doesn't set the FrameSlot#kind.